### PR TITLE
Introduce `sid::get_lower_bounds` and `sid::get_upper_bounds` 

### PR DESCRIPTION
**Pull request by anstaf**
- add `*_index` algorithms in `tuple_util`
- add `transform` and `for_each` for `hymap`
- introduce `stride_util` library
- introduce `make_contiguous` helper in `sid`
- use `make_contiguous` in cuda tmp implementation
- add `get_lower_bound`/`get_upper_bound` to the `SID` concept

### CODE REVIEW

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

File: include/gridtools/stencil_composition/sid/concept.hpp [hunk 1]
```diff
@@ -125,6 +125,8 @@
  *
  *  - Stride sid::get_stride<I>(Strides)
  *
+ * TODO(anstaf): add lower/upper bounds stuff when the API will be finally settled.
```
Comment (fthaler) [inline include/gridtools/stencil_composition/sid/concept.hpp]: Still true?

File: include/gridtools/stencil_composition/sid/synthetic.hpp [hunk 2]
```diff
@@ -73,12 +96,14 @@ namespace gridtools {
...
-                GT_CONSTEXPR synthetic<unique_mixin<Property, T>> set() const &&noexcept {
+                GT_CONSTEXPR synthetic<unique_mixin<Property, T>> set(
+                    property_constant<Property> = {}, meta::lazy::id<T> = {}) const &&noexcept {
                     return synthetic{};
                 }
 
                 template <property Property, class T>
-                GT_CONSTEXPR synthetic<unique_mixin<Property, decay_t<T>>> set(T &&val) const &&noexcept {
+                GT_CONSTEXPR synthetic<unique_mixin<Property, decay_t<T>>> set(
+                    T &&val, property_constant<Property> = {}) const &&noexcept {
```
Comment (fthaler) [inline include/gridtools/stencil_composition/sid/synthetic.hpp]: I find it quite confusing that for some properties, the `property_constant` comes first and for others it comes second. I guess this is due to the default values, but I think a uniform signature might be more important here.

File: include/gridtools/stencil_composition/sid/concept.hpp (hunk 1 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/sid/concept.hpp]: sorry, I meant to add documentation on it.

File: include/gridtools/stencil_composition/sid/synthetic.hpp (hunk 2 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/sid/synthetic.hpp]: Is it better now?

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
do perftest

File: include/gridtools/stencil_composition/sid/synthetic.hpp (hunk 2 above)
Comment (fthaler) [reply include/gridtools/stencil_composition/sid/synthetic.hpp]: Yes, I like it better that way!

**Pull request merged**